src/modules/land_detector/MulticopterLandDetector.cpp

```diff
@@ -131,115 +143,147 @@ void MulticopterLandDetector::_update_params()
+	}
+
+
+	// low thrust: 30% of throttle range between min and hover
+	const float sys_low_throttle = _params.minThrottle + (_params.hoverThrottle - _params.minThrottle) * 0.3f;
+
+	bool ground_contact = (_actuator_controls_throttle <= sys_low_throttle) && !vertical_movement;
 
 	// if we have a valid velocity setpoint and the vehicle is demanded to go down but no vertical movement present,
 	// we then can assume that the vehicle hit ground
-	_in_descend = _is_climb_rate_enabled()
-		      && (_vehicle_local_position_setpoint.vz >= land_speed_threshold);
-	bool hit_ground = _in_descend && !vertical_movement;
+	if (_flag_control_climb_rate_enabled) {
+		vehicle_local_position_setpoint_s vehicle_local_position_setpoint;
+
+		if (_vehicle_local_position_setpoint_sub.update(&vehicle_local_position_setpoint)) {
+			// setpoints can briefly be NAN to signal resets, TODO: fix in multicopter position controller
+			const bool descend_vel_sp = PX4_ISFINITE(vehicle_local_position_setpoint.vz)
+						    && (vehicle_local_position_setpoint.vz >= land_speed_threshold);
+
+			const bool descend_acc_sp = PX4_ISFINITE(vehicle_local_position_setpoint.acceleration[2])
+						    && (vehicle_local_position_setpoint.acceleration[2] >= 100.f);
+
+			_in_descend = descend_vel_sp || descend_acc_sp;
+		}
+
+		// ground contact requires commanded descent until landed
+		if (!_maybe_landed_hysteresis.get_state() && !_landed_hysteresis.get_state()) {
+			ground_contact &= _in_descend;
+		}
+
+	} else {
+		_in_descend = false;
+	}
+
+
+	// When not armed, consider to have ground-contact
+	if (!_armed) {
+		return true;
+	}
+
 
 	// TODO: we need an accelerometer based check for vertical movement for flying without GPS
-	return (_has_low_thrust() || hit_ground) && (!_horizontal_movement || !_has_position_lock())
-	       && (!vertical_movement || !_has_altitude_lock());
+	return ground_contact && !_horizontal_movement && !vertical_movement;
 }
 
 bool MulticopterLandDetector::_get_maybe_landed_state()
 {
 	// When not armed, consider to be maybe-landed
-	if (!_actuator_armed.armed) {
+	if (!_armed) {
 		return true;
 	}
 
-	if (_has_minimal_thrust()) {
-		if (_min_trust_start == 0) {
-			_min_trust_start = hrt_absolute_time();
+
+	// minimal throttle: initially 10% of throttle range between min and hover
+	float sys_min_throttle = _params.minThrottle + (_params.hoverThrottle - _params.minThrottle) * 0.1f;
+
+	// Determine the system min throttle based on flight mode
+	if (!_flag_control_climb_rate_enabled) {
+		sys_min_throttle = (_params.minManThrottle + 0.01f);
+	}
+
+	// Check if thrust output is less than the minimum auto throttle param.
+	if (_actuator_controls_throttle <= sys_min_throttle) {
+		if (_min_thrust_start == 0) {
+			_min_thrust_start = hrt_absolute_time();
 		}
 
 	} else {
-		_min_trust_start = 0;
+		_min_thrust_start = 0;
+		return false;
 	}
 
-	float landThresholdFactor = 1.0f;
+
+	if (_freefall_hysteresis.get_state()) {
```

Comment:
An observation: This will probably trigger only after the maybe landed hysteresis already expired since `MAYBE_LAND_DETECTOR_TRIGGER_TIME_US` is longer than `FREEFALL_TRIGGER_TIME_US` and freefall will probably occur when the thrust is cut with ground contact. But it will be before land detection since that adds enough time.

Reply:
It seemed harmless and I was imagining the case where you have min throttle for at least 8 seconds. So a very long freefall with minimal rotation... Not exactly a major concern, so I could go either way with this one.

Reply:
No it's fine, just wanted to point it out. Actually you're right it makes most sense in the case where it's combined with the 8 second low thrust and potentially no other way to detect if it's falling 👍